### ISSUE TITLE
big difference between numpy.matmul and torch.matmul

### ISSUE THREAD

**Issue by atomichighfive** [pytorch/pytorch#17678]
## 🐛 Bug
`numpy.matmul` and `torch.matmul` (after converting from numpy using `torch.from_numpy`) produces different results. The maximum absolute error is of the order 10^-4.

## To Reproduce

1. Download [this](https://drive.google.com/open?id=1PrH9GYP0o0549Degpj6UDbhA5Bjdo2OI) 1005001 by 20 (76.7 MiB) sample matrix.
2. Run code sample below with `matrix.npy` in working directory.

```
import numpy as np
import torch

torch.set_printoptions(precision=16)
np.set_printoptions(precision=16)

testJ = np.load('matrix.npy')
t_testJ = torch.from_numpy(testJ)

testPhi = np.matmul(testJ.T, testJ)
t_testPhi = torch.matmul(torch.t(t_testJ), t_testJ)

error = np.abs(testPhi - t_testPhi.data.numpy()).max()
t_error = torch.abs(torch.from_numpy(testPhi) - t_testPhi).max()

#print('Numpy matrix:')
#print(testPhi)
#print('Torch matrix:')
#print(t_testPhi)

print('torch version: %s\nnumpy version: %s\n' % (torch.__version__, np.__version__))

print('testJ type: %s\nt_testJ type: %s\ntestPhi type: %s\nt_testPhi type: %s\n' % (testJ.dtype, t_testJ.dtype, testPhi.dtype, t_testPhi.dtype))

print('Max absolute difference as numpy: %0.16f' % error)
print('Max absolute difference as torch: %0.16f' % t_error)
```

## Expected behavior
The maximum absolute difference between the result of the two different `matmul` operations comes out big (on the order 10^-4).

Running the reproduction code on my system produces:
```
torch version: 1.0.1.post2
numpy version: 1.16.2

testJ type: float32
t_testJ type: torch.float32
testPhi type: float32
t_testPhi type: torch.float32

Max absolute difference as numpy: 0.0002899169921875
Max absolute difference as torch: 0.0002899169921875
``` 

## Environment
Script output:
```
PyTorch version: 1.0.1.post2
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Manjaro Linux
GCC version: (GCC) 8.2.1 20181127
CMake version: Could not collect

Python version: 3.6
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy==1.14.3
[pip3] numpydoc==0.8.0
[conda] mkl                       2019.1                      144    anaconda
[conda] pytorch                   1.0.1           py3.6_cuda10.0.130_cudnn7.4.2_2    pytorch
[conda] torchvision               0.2.2                      py_3    pytorch
```
- pytorch was installed using: anaconda3
- exact python version is: Python 3.6.7

## Additional context

Not for now. Will edit if asked.

**Comment by albanD** [comment]
Hi,

Welcome to the wonderful world of float operations.
tldr: this is expected behavior because float operations are inexact.
The order in which the ops are done will change the result and if you accumulate a large number of values (millions in your case), this difference will grow quite a lot.
The idea for matmul is the same as for sum below.

Here is a small code sample where the exact same array is summed but in different order. You can check that changing the number of threads will give different results as well.

```python
import torch
import numpy as np


# Moving to double should reduce the difference to something very small ~1e-10
used_type = torch.float
# used_type = torch.double

# Changing the number of threads will potentially change the order of execution
# And thus change all the results
num_threads = 12

torch.manual_seed(1)
torch.set_num_threads(num_threads)
big_size = 1000000
a = torch.randn((big_size, 20), dtype=used_type)
print("Running with type {} and {} threads".format(used_type, num_threads))

sum_a = a.sum().item()
sum_a_np = np.sum(a.numpy())
print("original diff: ", sum_a - sum_a_np)

print("diffs: torch vs torch \t| torch vs np   \t| np vs np")
for _ in range(20):
    idx = torch.randperm(big_size)
    shuffled_a = a.index_select(0, idx)
    new_sum_a = shuffled_a.sum().item()
    new_sum_a_np = np.sum(shuffled_a.numpy())
    print("diffs: {} \t| {} \t| {}".format(sum_a - new_sum_a, new_sum_a - new_sum_a_np, sum_a_np - new_sum_a_np))
```
